### DISCUSSION TITLE
v1.28.0 requires the NET_BIND_SERVICE capability

### DISCUSSION THREAD

**Discussion by spantaleev** [dani-garcia/vaultwarden#3387] [Q&A]
I'm launching a Vaultwarden container like this:

```sh
docker run \
-it \
--rm \
--name=my-vaultwarden \
--log-driver=none \
--user=12345:12345 \
--cap-drop=ALL \
--read-only \
-e ROCKET_PORT=8080 \
--tmpfs=/tmp:rw,noexec,nosuid,size=128m \
--tmpfs=/data:rw,noexec,nosuid,size=128m \
--tmpfs=/cache:rw,noexec,nosuid,size=128m \
--tmpfs=/ephemeral:rw,noexec,nosuid,size=128m \
docker.io/vaultwarden/server:1.28.0-alpine
```

And it leads to an error like this:

> /start.sh: exec: line 25: /vaultwarden: Operation not permitted

Adding `--cap-add=NET_BIND_SERVICE \` **after** `--cap-drop=ALL \` fixes the problem and lets the process proceed further.

When using 1.27.0 instead of 1.28.0, the `NET_BIND_SERVICE` capability was not required. This can be tried by changing the version in the original command above and not adding `NET_BIND_SERVICE`.

I've explicitly configured `ROCKET_PORT` to use `8080` in an effort to escape needing `NET_BIND_SERVICE` in this container.

------

This seems to be caused by https://github.com/dani-garcia/vaultwarden/pull/3170

By adding the `NET_BIND_SERVICE` to the executable, it can no longer be used without having this privilege, not matter which ports you will need later.

Building my own container image that removes the capabilities added to `/vaultwarden` fixes the problem:

```Dockerfile
FROM docker.io/vaultwarden/server:1.28.0-alpine

RUN apk add --no-cache libcap && setcap cap_net_bind_service=-ep /vaultwarden
```

Seems like #3170 is only usable for niche situations with old Docker versions and is causing trouble otherwise. I suppose it's better to revert it.

**Comment by jjlin** [comment]
#3170 is intended to simplify running as a non-root user inside the container, particularly for less advanced users. While Docker 20.10.0 and above will automatically allow even non-root processes to bind to ports <1024, there are still users on older Docker versions, as well as users of other tools like Podman, which doesn't have this allowance AFAIK.

Is there a specific security concern with allowing `NET_BIND_SERVICE` in your use case? Arguably, the way you do things is also pretty niche.

**Comment by spantaleev** [comment]
I don't mind running it with `NET_BIND_SERVICE`, but given that it doesn't bind to such ports, it seems unnecessary and surprising. It took me quite a while to figure out why I was getting `Operation not permitted` when launching the process.

It seems like this convenience for less advanced users comes with at a slight security cost for others and a surprising complexity for people trying to drop all privileges for the container.

I believe the defaults should be adjusted so that the container does not use privileged ports by default and no such capability shenanings are necessary. Exposing a port `80` is done by Docker itself when publishing the ports, so there's really no need to bind to port `80` inside the container and force the need for the `NET_BIND_SERVICE` capability.

**Comment by jjlin** [comment]
Hindsight is 20/20 of course, but changing the default listen port in the Docker image at this point would likely break things for the vast majority of existing users.

**Comment by Omar007** [comment]
Tried updating and ran into this problem as well. Very unsure why this change to require a hard-coded capability at all times was made. Any user that has this running and configured their system properly has to already be in one of three categories; container configured with the capability added because it is running as non-root on a privileged port, configured to run as root on w/e port, or configured to run non-root on a non-privileged port. It wouldn't currently be running in the first place otherwise.  
So who is this helping or supposed to help? It can really only affect new users who are now forced to configure their systems/containers in a way not within the realm of one of the three proper deployment configurations  :confused: :thinking:   

I can understand not wanting to change the default port and this shouldn't be needed but the current change does not seem like the proper action either.  
This seems to be something that should probably just have been a documentation update instead of encouraging bad practices (enabling/requiring capabilities/privileges not actually needed).

**Comment by jjlin** [comment]
It's intended to help anyone who is using the default configuration (running as root on port 80) to transition more easily to running as non-root (still on the default port 80) simply by adding an additional `--user` option.

**Reply by Omar007** [reply]
The current solution then really sounds like an X-Y problem style solution. If someone wanted to make that transition, the docs could've been updated with a chapter/section about transitioning from root to non-root on Docker :thinking: 

If you're starting out fresh, the current docs already give use-cases on how to start as non-root (using a different port in this case). This section could easily have been expanded with a Docker as non-root on root ports use-case if that is the real problem.

**Reply by spantaleev** [reply]
So it's helping people following bad practices, by forcing all of us (even those who aren't following bad practices) into suddenly following bad practices and having to work around breaking changes.

As far as compatibility goes, this seems to be breaking compatilibity for a few of us at least, so changing the default port (even if such extreme measure are taken) doesn't seem that much worse than the current situation.

**Reply by Omar007** [reply]
> As far as compatibility goes, this seems to be breaking compatilibity for a few of us at least, so changing the default port (even if such extreme measure are taken) doesn't seem that much worse than the current situation.

Arguably true but I do suggest/propose sticking to a solution that is both compatible with existing installations as well as new ones. Just making a different breaking change because there is a change that broke things does not seem right to me haha.  
Or if we are making breaking changes, big version bump and migration notes/docs I suppose. But in that case, better make the proper breaking change... :sweat_smile:

**Comment by Omar007** [comment]
Just spotted another one in the 1.28 discussion affected by this that I had missed yesterday (only looked through issues, would've saved me up to an hour or so of searching, going though diffs and debugging D:  ); https://github.com/dani-garcia/vaultwarden/discussions/3374#discussioncomment-5441347

**Comment by xduugu** [comment]
Really unfortunate this change. Wouldn't it be possible to simply set an unprivileged port as default for ROCKET_PORT?

**Reply by BlackDex** [reply]
That would break all existing installations, so that is not an option actually. Removing this cap, might be from my point of view if this really causes a lot of issues.

**Reply by xduugu** [reply]
Yeah, you're right. But why was it changed at all? Only to not have to set an environment variable? In my opinion, setting ROCKET_PORT as described on the wiki page before, was the much cleaner solution and if you really want to expose vaultwarden on port 80, you can map it to this port using docker's `-p`/`--publish` option.

https://github.com/dani-garcia/vaultwarden/wiki/Hardening-Guide/_compare/a7cde75df636aabec369dc12383ea18af9f5c6f2...672e5db240f087854e49b0d44de0ec6eaa3b7b6c

**Reply by BlackDex** [reply]
The port default in the image is currently set to 80. So nothing changed there. What has changed, is making it easier for new users, running these images as non-root user. Thats it, nothing more nothing less.

**Reply by xduugu** [reply]
Are you referring to the [wiki page](https://github.com/dani-garcia/vaultwarden/wiki/Hardening-Guide#run-as-a-non-root-user) or the [docker command in the README](https://github.com/dani-garcia/vaultwarden/blob/525e6bb65a6926e0f9de3fc5dafd5c5b63981f9f/README.md?plain=1#L41)?

How about adding `--sysctl net.ipv4.ip_unprivileged_port_start=0` to the docker command or the corresponding lines to the compose file? With this option, vaultwarden can run on port 80 without root or any capability.

By the way, if somebody wants to remove the capability from the image, you can use this (alpine) Dockerfile. Unfortunately, the `setcap` command is not available in the image and needs to be installed first.

```Dockerfile
FROM docker.io/vaultwarden/server:alpine

RUN apk add --no-cache --virtual .build-deps libcap-utils \
  && setcap -r /vaultwarden \
  && apk del .build-deps
```

**Comment by spantaleev** [comment]
I'm sorry, but this is turning into some sort of joke and it needs to be called out.

I'm curious:

1. Who **exactly** is #3170 helping? I don't see anyone asking for this
2. You keep quoting backward compatibility but you obviously broke it in a bad way, judging by the fact that at least a bunch of us are here, having had to work around it and having wasted hours on it
3. Why are you catering to a bunch of likely non-existent outdated setups with #3170 and not to your up to date and future users?
4. Why are you shooting yourself in the foot catering to hypothetical users with #3170 and preventing yourself from encouraging actually good practices (higher port number for `ROCKET_PORT` and running with privileges dropped) like us here have been following? Now that #3170 has landed, you prevent yourself from doing the right thing in the future
5. What do you think the handling of this situation tells people about Vaultwarden (a **security** related product) and how this project is managed and who it caters to? Catering to hypothetical users with ancient setups who can't be bothered to tweak or update them, instead of actually encouraging best security practices
6. Why can't you just say "I fucked up, let's at least revert it"?

**Reply by BlackDex** [reply]
First, Let's keep it nice.

Second, the PR you are referring to had not been merged for **18 days**, and nobody commented on this for being wrong or a bad idea.

Third, this change has been part for more then a month in testing, and again nobody complained.

I'm not saying we should not do anything right now. But i find it a bit disappointing that words like above are used for something done with good intentions and nobody complained back then. In hindsight everything can be a good or bad decision.

Also, this is an open source project, we do except well written PR's 😉.

**Comment by BlackDex** [comment]
I have created a PR to revert the change #3403

**Reply by xduugu** [reply]
Thanks! v1.28.1 works again without additional capabilities.
